File: 0127-word-ladder/0127-word-ladder.java
class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        if (!wordList.contains(endWord)) return 0;

        Set<String> wordSet = new HashSet<>(wordList);
        Set<String> visited = new HashSet<>();
        Queue<Pair> queue = new LinkedList<>();
        visited.add(beginWord);
        queue.offer(new Pair(beginWord, 1));

        while (!queue.isEmpty()) {
            Pair cur = queue.poll();
            String word = cur.word;
            int step = cur.step;

            if (word.equals(endWord)) return step;

            for (int i = 0; i < word.length(); i++) {
                for (char c = 'a'; c <= 'z'; c++) {
                    StringBuilder sb = new StringBuilder(word);
                    sb.setCharAt(i, c);
                    String newWord = sb.toString();

                    if (wordSet.contains(newWord) && !visited.contains(newWord)) {
                        visited.add(newWord);
                        queue.offer(new Pair(newWord, step + 1));
                    }
                }
            }
        }

        return 0;
    }

    private class Pair {
        String word;
        int step;

        Pair(String word, int step) {
            this.word = word;
            this.step = step;
        }
    }
}

File: 0127-word-ladder/0127-word-ladder.kt
class Solution {
    fun ladderLength(beginWord: String, endWord: String, wordList: List<String>): Int {
        if (endWord !in wordList) return 0

        val wordSet = wordList.toSet()
        val visited = mutableSetOf<String>()
        val queue = ArrayDeque<Pair<String, Int>>()
        
        visited.add(beginWord)
        queue.add(beginWord to 1)

        while (queue.isNotEmpty()) {
            val (word, step) = queue.removeFirst()

            if (word == endWord) return step

            for (i in word.indices) {
                for (c in 'a'..'z') {
                    val newWord = buildString {
                        append(word.substring(0, i))
                        append(c)
                        append(word.substring(i + 1))
                    }

                    if (newWord in wordSet && newWord !in visited) {
                        visited.add(newWord)
                        queue.add(newWord to step + 1)
                    }
                }
            }
        }

        return 0
    }
}